pkg/streamingpromql/operators/selectors/range_vector_selector_test.go

```diff
@@ -0,0 +1,297 @@
+			step:          time.Minute,
+			rangeDuration: time.Minute * 2,
+			anchored:      false,
+			steps:         4,
+		},
+		// anchored - set the step duration > range duration
+		"anchored - step > range": {
+			step:          time.Minute,
+			rangeDuration: time.Second * 30,
+			anchored:      true,
+			steps:         4,
+		},
+		// smoothed - set the step duration > range duration
+		"smoothed - step > range": {
+			step:          time.Minute,
+			rangeDuration: time.Second * 30,
+			anchored:      false,
+			smoothed:      true,
+			steps:         4,
+		},
+		// anchored - set the step duration = range duration
+		"anchored - step = range": {
+			step:          time.Second * 30,
+			rangeDuration: time.Second * 30,
+			anchored:      true,
+			steps:         7,
+		},
+		// smoothed - set the step duration = range duration
+		"smoothed - step = range": {
+			step:          time.Second * 30,
+			rangeDuration: time.Second * 30,
+			anchored:      false,
+			smoothed:      true,
+			steps:         7,
+		},
+	}
+
+	for name, tc := range tests {
+		t.Run(name, func(t *testing.T) {
+			ctx := context.Background()
+			mc := limiter.NewMemoryConsumptionTracker(ctx, 0, nil, "")
+			tr := types.NewRangeQueryTimeRange(timeZero, timeZero.Add(3*time.Minute), tc.step)
+
+			selector := &Selector{
+				Queryable:                storage,
+				TimeRange:                tr,
+				Matchers:                 []types.Matcher{{Type: labels.MatchEqual, Name: model.MetricNameLabel, Value: "metric"}},
+				LookbackDelta:            1 * time.Minute,
+				Range:                    tc.rangeDuration,
+				Anchored:                 tc.anchored,
+				Smoothed:                 tc.smoothed,
+				MemoryConsumptionTracker: mc,
+			}
+
+			rv := NewRangeVectorSelector(selector, mc, types.NewQueryStats(), tc.anchored, tc.smoothed)
+
+			metadata, err := rv.SeriesMetadata(ctx, nil)
+			require.NoError(t, err)
+			assert.Len(t, metadata, 2)
+
+			for range metadata {
+				require.NoError(t, rv.NextSeries(ctx))
+
+				// This is a range query. The start goes back the given range duration
+				start := timestamp.Time(tr.StartT).Add(0 - tc.rangeDuration)
+				end := timestamp.Time(tr.StartT)
+
+				// Iterate through all the steps, validating that we increment at the given step value
+				for range tc.steps {
+					stepData, err := rv.NextStepSamples(ctx)
+					require.NoError(t, err)
+					require.NotNil(t, stepData)
+
+					require.Equal(t, timestamp.FromTime(start), stepData.RangeStart)
+					require.Equal(t, timestamp.FromTime(end), stepData.RangeEnd)
+
+					start = start.Add(tc.step)
+					end = end.Add(tc.step)
+				}
+
+				_, err := rv.NextStepSamples(ctx)
+				require.Equal(t, types.EOS, err)
+			}
+		})
+	}
+}
+
+func TestRangeVectorSelectorSyntheticPoints(t *testing.T) {
+
+	timeZero := time.Unix(0, 0)
+
+	tests := map[string]struct {
+		data            string
+		ts              time.Time
+		expected        []promql.FPoint
+		hasSmoothedHead bool
+		hasSmoothedTail bool
```

Comment:
[nit] This might make it clearer that these are expectations, not test input data.

```suggestion
		expectedHasSmoothedHead bool
		expectedHasSmoothedTail bool
```